theano/scalar/basic.py

```diff
@@ -277,10 +277,35 @@ def convert(x, dtype=None):
     return x_
 
 
-def constant(x):
-    x = convert(x)
-    assert x.ndim == 0
-    return ScalarConstant(get_scalar_type(str(x.dtype)), x)
+def constant(x, name=None, dtype=None):
+    # pass through numpy scalars, since they are already typed on
+    # purpose typically.
+    if dtype is not None:
+        x = theano._asarray(x, dtype=dtype)
+    if hasattr(x, 'dtype'):
+        assert x.ndim == 0
+        return ScalarConstant(get_scalar_type(str(x.dtype)), x, name=name)
+    if isinstance(x, builtin_float):
+        for dtype in ['float32', 'float64']:
+            x_ = theano._asarray(x, dtype=dtype)
+            if numpy.all(x == x_):
+                break
+            x_ = None
+        assert x_ is not None
+        return ScalarConstant(get_scalar_type(str(x_.dtype)), x, name=name)
+    if isinstance(x, builtin_int):
+        for dtype in ['int8', 'int16', 'int32', 'int64']:
+            x_ = theano._asarray(x, dtype=dtype)
+            if numpy.all(x == x_):
+                break
+            x_ = None
+        assert x_ is not None
+        return ScalarConstant(get_scalar_type(str(x_.dtype)), x, name=name)
+    if isinstance(x, builtin_complex):
+        # TODO: We have added the complex type, so this should be tested
+        raise NotImplementedError()
+    raise TypeError(x)
```

Comment:
Good catch. I didn't do this in purpose. It come from a rebase. convert wasn't there when this was done. I'll update this.